Add an export that allows for completely out-of-source use of Serac

joshessman-llnl opened the pull request:
This is mostly a duplication of the approach used by Axom to "manually' re-import serac's dependencies, and was tested by building + linking the current serac driver in a completely different directory with the following minimal `CMakeLists.txt` (path to config file was hardcoded for simplicity):
```cmake
cmake_minimum_required(VERSION 3.8)
project(test LANGUAGES C CXX)

include(CMakeFindDependencyMacro)
find_dependency(serac REQUIRED NO_DEFAULT_PATH PATHS "../../lib/cmake")

add_executable(s_test test.cpp)
set_property(TARGET s_test PROPERTY CXX_STANDARD 17)
target_link_libraries(s_test cli11 serac::serac)

```

Resolves #154

joshessman-llnl committed:
feat: added config file that manually re-imports dependencies for completely out-of-source builds

joshessman-llnl commented:
CMake 3.13 introduced a change affecting when `install` can be called on a target

joshessman-llnl committed:
fix: move install commands into subdirs for compatibility with CMake < 3.13

samuelpmishLLNL commented:
This looks very promising! Could we put the serac targets behind a namespace? Or just group the serac targets into a single target "serac" to make the usage more like

`target_link_libraries(s_test cli11 serac)`
or
`target_link_libraries(s_test cli11 serac::coefficients serac::solvers)`

joshessman-llnl committed:
feat: added new serac exported target that includes all serac components

joshessman-llnl committed:
Merge branch 'develop' into feature/essman/export-cmake-target

joshessman-llnl commented:
> 
> 
> This looks very promising! Could we put the serac targets behind a namespace? Or just group the serac targets into a single target "serac" to make the usage more like
> 
> `target_link_libraries(s_test cli11 serac)`
> or
> `target_link_libraries(s_test cli11 serac::coefficients serac::solvers)`

I went ahead and added both in case the user wants to only link against a subset of the serac libraries, I've updated the sample `CMakeLists.txt` above to reflect that the user can just specify `serac::serac` if they want everything, or they can use your second snippet.

white238 commented on cmake/serac-config.cmake.in:
```diff
@@ -0,0 +1,103 @@
+# SPDX-License-Identifier: (BSD-3-Clause)
+#------------------------------------------------------------------------------
+
+cmake_minimum_required(VERSION 3.0 FATAL_ERROR)
```
BLT requires 3.8.  This also has side effects of loading CMake policies differently depending on what version you list here (even if you are actually using 3.14.0 or whatever).  I would keep it as close to what we actually require.  Not sure why axom chose 3.0 since they definitely dont work with 3.0

white238 commented:
Can you add an example project that uses our exported config and add a test to our PR's that tests against it so that it continues to work.  These config's have a habit of breaking.  It can be a basic smoke level test.

Axom has an example of this...

joshessman-llnl committed:
tests: add a test to make sure that external projects can use serac

joshessman-llnl committed:
Merge branch 'develop' into feature/essman/export-cmake-target

joshessman-llnl committed:
docs: add requirement that TPLs be built outside of the local repo

joshessman-llnl committed:
Merge branch 'develop' into feature/essman/export-cmake-target

joshessman-llnl committed:
Merge branch 'develop' into feature/essman/export-cmake-target

joshessman-llnl committed:
Merge branch 'develop' into feature/essman/export-cmake-target

joshessman-llnl committed:
Merge branch 'ci/essman/unify_scripts' into feature/essman/export-cmake-target

joshessman-llnl committed:
Merge branch 'develop' into feature/essman/export-cmake-target

salazardetroya commented on examples/cmake/serac.cpp:
```diff
@@ -0,0 +1,24 @@
+
+#include "common/common.hpp"
+
+int main(int argc, char* argv[])
```
What is the reason to have the example source files inside of the `cmake/` folder? I find it confusing since usually the `cmake/` folder contains `CMake` related files.

jamiebramwell replied:
I believe it's because this is an example of how to use the exported cmake project. @white238 how do other projects do this?

joshessman-llnl replied:
Yep, this was intended to follow the structure axom uses (https://github.com/LLNL/axom/tree/develop/src/examples).  I will rename this directory to `using-with-cmake` for clarity.

white238 replied:
the nice thing about having this in the repo is we can test it as part of our normal CI to make sure it works.  The renaming the directory will help with the clarity...

joshessman-llnl committed:
Merge branch 'develop' into feature/essman/export-cmake-target

jamiebramwell commented:
Does this need more work or is it ready for another review/merging?

joshessman-llnl committed:
fix: rename cmake example directory for clarity

fix: updated example directory in test

white238 commented on CMakeLists.txt:
```diff
@@ -70,10 +70,31 @@ if(ENABLE_TESTS)
+#------------------------------------------------------------------------------
+# Export Targets
+#------------------------------------------------------------------------------
+add_library(serac INTERFACE)
```
Did having two CMake targets named the same thing not cause an error? serac the executable and serac the library?

white238 commented on cmake/serac-config.cmake.in:
```diff
@@ -0,0 +1,102 @@
+  # mfem
+  if(SERAC_USE_MFEM)
+    set(SERAC_MFEM_DIR     "@MFEM_DIR@")
+    # Note: Targets not currently imported
```
Should we be calling the FindMFEM here or leaving it up to the host-code to handle this themselves?

joshessman-llnl replied:
Did you mean in the case where they'd want to link to MFEM seperately?

white238 replied:
Are we expecting users of this to be able to make mfem calls?

joshessman-llnl replied:
By linking to the `serac` targets the MFEM libs get pulled in, and the MFEM headers have `INTERFACE` visibility, so the example can just `#include "mfem.hpp"` and the rest should work.  I'll add some direct MFEM calls to the `serac_example.cpp` for clarity.

joshessman-llnl committed:
fix: update path to example src file

white238 commented on src/drivers/CMakeLists.txt:
```diff
@@ -4,11 +4,11 @@
 # SPDX-License-Identifier: (BSD-3-Clause)
 
-blt_add_executable( NAME       serac
+blt_add_executable( NAME       serac_driver
```
Should probably go through the entire PR before making comments.

joshessman-llnl committed:
feat: print axom/mfem version info in example to confirm includes/links work

joshessman-llnl committed:
Merge branch 'develop' into feature/essman/export-cmake-target

jamiebramwell committed:
Merge branch 'develop' into feature/essman/export-cmake-target

jamiebramwell commented:
LGTM

jamiebramwell commented:
Looks good! @white238 , do you have any outstanding issues with this PR?

white238 commented:
We will see when someone tries to do this... it usually requires several go arounds to catch all the things that were missed.
